fulltext/routes.py

```diff
@@ -0,0 +1,135 @@
+"""Provides the blueprint for the fulltext API."""
+
+from typing import Optional, Callable, Any, List
+from flask import request, Blueprint, Response
+from werkzeug.exceptions import NotAcceptable, BadRequest, NotFound
+from flask.json import jsonify
+from arxiv import status
+from arxiv.users.domain import Session, Scope
+from arxiv.users.auth import scopes
+from arxiv.users.auth.decorators import scoped
+from arxiv.base import logging
+from fulltext import controllers
+from .domain import SupportedBuckets, SupportedFormats
+
+logger = logging.getLogger(__name__)
+
+ARXIV_PREFIX = '/<id_type>/<arxiv:identifier>'
+SUBMISSION_PREFIX = '/<id_type>/<source:identifier>'
+
+blueprint = Blueprint('fulltext', __name__, url_prefix='')
+
+Authorizer = Callable[[str, Optional[str]], bool]
+
+
+def make_authorizer(scope: Scope) -> Authorizer:
+    """Make an authorizer function for injection into a controller."""
+    def inner(identifier: str, owner_id: Optional[str]) -> bool:
+        """Check whether the session is authorized for a specific resource."""
+        logger.debug('Authorize for %s owned by %s', identifier, owner_id)
+        logger.debug('Client user id is %s', request.auth.user.user_id)
+        try:
+            source_id, checksum = identifier.split('/', 1)
+        except ValueError as e:
+            logger.debug('Bad identifier? %s', e)
+            raise NotFound('Unsupported identifier') from e
+        return (request.auth.is_authorized(scope, source_id)
+                or (request.auth.user
+                    and str(request.auth.user.user_id) == owner_id))
+    return inner
+
+
+def resource_id(id_type: str, identifier: str, *args: Any, **kw: Any) -> str:
+    """Get the resource ID for an endpoint."""
+    if id_type == SupportedBuckets.SUBMISSION:
+        return identifier.split('/', 1)[0]
+    return identifier
+
+
+def best_match(available: List[str], default: str) -> str:
+    """Determine best content type given Accept header and available types."""
+    if 'Accept' not in request.headers:
+        return default
+    ctype: str = request.accept_mimetypes.best_match(available)
+    return ctype
+
+
+@blueprint.route('/status')
+def ok() -> tuple:
+    """Provide current integration status information for health checks."""
+    data, code, headers = controllers.service_status()
+    return jsonify(data), code, headers
+
+
+@blueprint.route(ARXIV_PREFIX, methods=['POST'])
+@blueprint.route(SUBMISSION_PREFIX, methods=['POST'])
+@scoped(scopes.CREATE_FULLTEXT, resource=resource_id)
+def start_extraction(id_type: str, identifier: str) -> tuple:
+    """Handle requests for fulltext extraction."""
+    force = request.args.get('force', False)
+    token = request.environ['token']
+
+    # Authorization is required to work with submissions.
+    authorizer: Optional[Authorizer] = None
+    if id_type == SupportedBuckets.SUBMISSION:
+        authorizer = make_authorizer(scopes.READ_COMPILE)
```

Comment:
When doing authorization, defaulting to restricted with some cases authorized for everyone, is safer than defaulting to authorized for everyone, with some cases restricted.

Reply:
Indeed. 

[L66](https://github.com/arXiv/arxiv-fulltext/pull/16/files/f17f1c714db42a33bca34467faf722b9d89a8c22#diff-eee66427890486dd43fdcedae6ba165bR66): the request must be authenticated, and be authorized for ``CREATE_FULLTEXT``. 
```python
@scoped(scopes.CREATE_FULLTEXT, resource=resource_id)
```

[L74-75](https://github.com/arXiv/arxiv-fulltext/pull/16/files/f17f1c714db42a33bca34467faf722b9d89a8c22#diff-eee66427890486dd43fdcedae6ba165bR74) is for the special case that we are handling text extraction for a submission. In that case, the request must also be specifically authorized to ``READ_COMPILE`` this resource and/or be the owner of the resource. I wrote it this way so that it is clear that it applies to the submission case.